### DISCUSSION TITLE
Suggestion to use `blender_bevy_toolkit` as our main game engine editor.

### DISCUSSION THREAD

**Discussion by Joe23232** [bevyengine/bevy#2825] [Ideas]
A lot of game engines such as Godot and Unity comes with a gui game editor.

I understand Bevy has a roadmap to design such a gui game engine editor, however already somebody has made an exporter addon for Blender which is [`blender_bevy_toolkit`](https://github.com/sdfgeoff/blender_bevy_toolkit) which exports to Bevy which is close to a gui game engine editor.

Rather than "reinventing the wheel" why not instead continue to support this addon and continue to improve it instead of making a whole different gui game engine editor for Bevy?

**Comment by Nilirad** [comment]
The whole point about making a game engine _is_ to reinvent the wheel, at least to an extent. While basing the editor on blender has its own advantages, there are other concerns: what about licensing, performance, customization, extensibility?

Also, blender is designed to be a 3D modeling software. Bevy instead needs a game engine editor with its own sets of features that blender will never support. The only useful things that blender can do out of the box for an editor is a world viewport, mouse picking and keyboard shortcuts (and maybe something else that I didn't think about). Everything else must be done from scratch, probably with many headaches caused by how the blender API is designed.

**Reply by Joe23232** [reply]
> Also, blender is designed to be a 3D modeling software. Bevy instead needs a game engine editor with its own sets of features that blender will never support

I thought blender is pretty customisable via Python scripts that the addons serve/do, I have seen addons that have drastically change Blender's userability so yes it is possible to heavily modify Blender's userinterface via addons mate :)

> Everything else must be done from scratch, probably with many headaches caused by how the blender API is designed.

_maybe_

Can I ask you is there like an offical alpha gui game editor for Bevy?

**Reply by bjorn3** [reply]
> Can I ask you is there like an offical alpha gui game editor for Bevy?

There is the unofficial [bevy_editor_pls](https://github.com/jakobhellermann/bevy_editor_pls) but something official is waiting on UI work AFAICT.

**Reply by alterae** [reply]
> I thought blender is pretty customisable via Python scripts that the addons serve/do, I have seen addons that have drastically change Blender's userability so yes it is possible to heavily modify Blender's userinterface via addons mate :)

While blender may be extensible, the setup required and the additional blender-specific features are not exactly the most user friendly way to go about it.  Plus, the world viewport is a fairly minimal part of the editor requirements.  The big thing is serialization and deserialization of scenes and code, which is something Blender is very much not built for.

While using blender is theoretically possible, it would require us to keep up with and work around their API, and would result in a substantial quantity of overhead.

And if we were to use blender we would not get the advantages of building our own editor (namely, tight integration and an incentive to improve the engine's UI capabilities in order to build the editor)

> Can I ask you is there like an offical alpha gui game editor for Bevy?

There is not.  I don't have the links to the relevant issues right now but look up "Editor-ready UI" in the issue tracker.

**Reply by Joe23232** [reply]
> There is the unofficial bevy_editor_pls but something official is waiting on UI work AFAICT.

ARe the going to build on top of that mate?

> While blender may be extensible, the setup required and the additional blender-specific features are not exactly the most user friendly way to go about it. Plus, the world viewport is a fairly minimal part of the editor requirements. The big thing is serialization and deserialization of scenes and code, which is something Blender is very much not built for.

Right I see mate.


> There is not. I don't have the links to the relevant issues right now but look up "Editor-ready UI" in the issue tracker.

Ok sure thanks mate :)

**Comment by afonsolage** [comment]
I don't think it's a good idea, since by creating our own editor we'll be able to extract all potential of Bevy as an engine. Also, the Bevy Editor is a good case to test our own crates, like UI, Assets and such.
